proposals/0330-collection-conditionals.md

```diff
@@ -36,13 +40,168 @@ The most notable use case for this is conditional inclusion of tests for the Swi
++ dictionary-literal-items → dictionary-literal-item ','(opt) | dictionary-literal-item ',' dictionary-literal-items | conditional-compilation-block dictionary-literal-items
+```
+
+If a `conditional-compilation-block` is inside `array-literal` or `dictionay-literal`, it can only contain `array-literal-items` or `dictionary-literal-items` respectively.
+
+### Array literal or dictionary literal
+
+Given a collection literal with conditional compilation directives in it, whether it is an array literal or dictionary literal is determined solely by the first occurence of the element in the source text. Regardless of whether it is in the active block or not, if there is a colon after the expression of the first element, the literal is treated as a dictionary literal; otherwise, it is treated as an array literal. Therefore, a literal that has a dictionary element in a block and an array element in another block will be treated as an error.
+
+```swift
+let literal = [
+#if CONDITION_1
+  key: value,
+#else
+  value, // error: expected ':' and a dictionary literal element
+#endif
+]
+```
+
+### Empty dictionary literal
+
+A dictionary literal can be empty depending on the conditional compilation directives. In the following case, if the `CONDITION`evaluates to `false`, this is considered an empty dictionary literal. 
+
+```swift
+let literal = [
+#if CONDITION
+  "foo": "bar",
+#endif
+]
+```
+
+You cannot and don't need to write `:` in a block  to emulate a empty dictionally litral  `[:]`.
+
+```swift
+let literal = [
+#if CONDITION
+  "foo": "bar",
+#else
+  : // error: expected key expression in dictionary literal
+#endif
+]
+```
+
+### Empty block body
+
+Each block in `#if ... #endif` can be empty.
+
+```swift
+let literal = [
+#if CONDITION_1
+  // Empty.
+#else
+  key1: value1,
+#endif
+]
+```
 
-As conditionals themselves and inactive elements are not included in the parser AST representation, a new data structure, the "Conditionals Map" is maintained on the `CollectionExpr` which is used to support features such as the AST dump and stripping conditionals from module interfaces. The syntax model of libSyntax also required minor modification.
+Even though the top `#if` body is empty, this is parsed as a valid dictionary literal regardless of the value of `CONDITION_1` because the first element that appears (i.e. `key1: value1,`) contains a colon.
+
+### Type inference
+
+Type infrerence for collection literals is done based on the parsed _active_ alements after evaluating the conditional compilation directives.
+
+```swift
+class Animal {}
+class Dog: Animal {}
+class Cat: Animal {}
+
+let array = [
+  Dog(),
+#if CONDITION_1
+  Cat(),
+#endif
+]
+```
+
+In this example, if `CONDITION_1` evaluates to `true`, this array is inferred `Array<Base>` because `Animal` the best common type between `Dog()` and `Cat()`, otherwise it's `Array<Dog>` because `Dog()` is the only active element.
```

Comment:
```suggestion
In this example, if `CONDITION_1` evaluates to `true`, this array is inferred `Array<Animal>` because `Animal` the best common type between `Dog()` and `Cat()`, otherwise it's `Array<Dog>` because `Dog()` is the only active element.
```